tests/output-dsl.nf

```diff
@@ -16,75 +16,95 @@
  */
 nextflow.preview.output = true
 
-params.save_foo = true
+params.save_bam_bai = false
 
-process align {
+process fastqc {
   input:
-  val(x)
+  val id
 
   output:
-  path("*.bam")
-  path("${x}.bai")
+  tuple val(id), path('*.fastqc.log')
 
   script:
   """
-  echo ${x} > ${x}.bam
-  echo ${x} | rev > ${x}.bai
+  echo ${id} > ${id}.fastqc.log
   """
 }
 
-process my_combine {
+process align {
   input:
-  path(bamfile)
-  path(baifile)
+  val id
 
   output:
-  path 'result.txt'
+  tuple val(id), path('*.bam')
+  tuple val(id), path('*.bai')
 
   script:
   """
-  cat $bamfile > result.txt
-  cat $baifile >> result.txt
+  echo ${id} > ${id}.bam
+  echo ${id} | rev > ${id}.bai
   """
 }
 
-process foo {
+process quant {
+  input:
+  val id
+
   output:
-  path 'xxx'
+  tuple val(id), path('quant')
 
   script:
   '''
-  mkdir xxx
-  touch xxx/A
-  touch xxx/B
-  touch xxx/C
+  mkdir quant
+  touch quant/cmd_info.json
+  touch quant/lib_format_counts.json
+  touch quant/quant.sf
   '''
 }
 
 workflow {
   main:
-  input = Channel.of('alpha','beta','delta')
-  align(input)
+  ids = Channel.of('alpha', 'beta', 'delta')
+  ch_fastqc = fastqc(ids)
+  (ch_bam, ch_bai) = align(ids)
+  ch_quant = quant(ids)
 
-  bams = align.out[0].toSortedList { bam -> bam.name }
-  bais = align.out[1].toSortedList { bai -> bai.name }
-  my_combine( bams, bais )
-  my_combine.out.view { it -> it.text }
-
-  foo()
+  ch_samples = ch_fastqc
+    .join(ch_bam)
+    .join(ch_bai)
+    .join(ch_quant)
+    .map { id, fastqc, bam, bai, quant ->
+      [
+        id: id,
+        fastqc: fastqc,
+        bam: params.save_bam_bai ? bam : null,
+        bai: params.save_bam_bai ? bai : null,
+        quant: quant
+      ]
+    }
 
   publish:
-  align.out       >> 'data'
-  my_combine.out  >> 'more/data'
-  foo.out         >> (params.save_foo ? 'data' : null)
+  ch_samples >> 'samples'
 }
 
 output {
-  'data' {
-    path { val -> { file -> file } }
+  samples {
+    path { sample ->
+      def dirs = [
+        'bam': 'align',
+        'bai': 'align',
+        'log': 'fastqc'
+      ]
+      return { filename ->
+        def ext = filename.tokenize('.').last()
+        def dir = dirs[ext]
+        dir != null
+          ? "${dir}/${filename}"
+          : "${filename}/${sample.id}"
+      }
+    }
```

Comment:
@pditommaso regarding your main comment

The purpose of the dynamic path directive is to map a data structure, like a record, to a directory tree, so that the directory structure can be independent of the index file structure. This decoupling is crucial.

I agree that it is ugly. I would like to find a more declarative syntax for this that is sufficiently flexible. I have a few ideas cooking. But this nested closure is the way to do it in the second preview. So right now the goal is to fix the bugs so that it behaves as intended. The sooner we merge these fixes, the sooner I can start working on a third preview / final version.

Reply:
Still struggling my head about this snipper. If i'm understanding correctly `sample` is a map holding a the `id` and other file paths. I guess the closure in then invoked on all map entries that are Path. Is that correct?

Reply:
yes

Reply:
Another idea based on our discussion today:

```groovy
    path { sample ->
      sample.fastqc >> 'log/'
      sample.bam >> 'align/'
      sample.bai >> 'align/'
      sample.quant >> "quant/${sample.id}"
    }
```

More like a builder DSL, but still requires a closure to access the record members.